cirq-core/cirq/transformers/routing/visualize_routed_circuit.py

```diff
@@ -0,0 +1,70 @@
+# Copyright 2022 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Dict, Optional, Tuple, TYPE_CHECKING
+from cirq import circuits, ops
+
+if TYPE_CHECKING:
+    import cirq
+
+
+class _SwapPrintGate(ops.Gate):
+    """A gate that displays the string representation of each qubits on the circuit."""
+
+    def __init__(self, qubits: Tuple[Tuple['cirq.Qid', 'cirq.Qid'], ...]) -> None:
+        self.qubits = qubits
+
+    def num_qubits(self):
+        return len(self.qubits)
+
+    def _circuit_diagram_info_(self, args: 'cirq.CircuitDiagramInfoArgs') -> Tuple[str, ...]:
+        return tuple(f'{str(q[1])}' for q in self.qubits)
+
+
+def routed_circuit_with_mapping(
+    routed_circuit: 'cirq.AbstractCircuit',
+    initial_map: Optional[Dict['cirq.Qid', 'cirq.Qid']] = None,
+) -> 'cirq.AbstractCircuit':
+    """Returns the same circuits with information about the permutation of qubits after each swap.
+
+    Args:
+        routed_circuit: a routed circuit that potentially has inserted swaps tagged with a
+            RoutingSwapTag.
+        initial_map: the initial mapping from logical to physical qubits. If this is not specified
+            then the identity mapping of the qubits in routed_circuit will be used as initial_map.
+    """
+    all_qubits = sorted(routed_circuit.all_qubits())
+    qdict = {q: q for q in all_qubits}
+    if initial_map is None:
+        initial_map = qdict.copy()
+    inverse_map = {v: k for k, v in initial_map.items()}
+
+    ret_circuit = circuits.Circuit(
+        _SwapPrintGate(tuple(zip(qdict.values(), [inverse_map[x] for x in qdict.values()]))).on(
+            *all_qubits
+        )
+    )
+    for m in routed_circuit:
+        # Find the mapping at after this moment
+        for op in m:
+            if ops.RoutingSwapTag() in op.tags:
+                q1, q2 = op.qubits
+                qdict[q1], qdict[q2] = qdict[q2], qdict[q1]
+        ret_circuit.append(m)
+        ret_circuit.append(
```

Comment:
This injects qubit positions after _every_ moment. Is this intended, or should positions only be injected after swaps which affect them?

Reply:
I think originally the plan was to inject qubit position info after every swap, not every moment. but now I am wondering which is more readable. Any thoughts on this? I'll soon push a commit with the former.

Reply:
My preference would be only after swaps, just to keep minimize bloat. There's an argument to be made that every moment ensures there's always a nearby reference for which qubit is which, but that feels less pressing to me (and likely a non-issue for even moderately complex circuits)